README.md
# discord-uptime
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)

Discord Uptime is a Discord bot that allows you to monitor the uptime of services using ICMP ping, tcp and http requests.
There are also commands available to make manual requests. Built using discord.py, aioping and aiohttp.

## Installation
**Requires Python 3.8+**

Install dependencies:

`pip install -r requirements.txt`

Bot setup (Rename config.example.json and edit the default values):
* `token` - Discord bot token
* `prefix` - Discord bot prefix
* `activity_type` - Activity type for bot status (must be one of `playing, streaming, listening, watching`)
* `activity_name` - Text for bot status
* `disable_help` - If true, the default help command will be disabled
* `hide_ips` - If true, any IP addresses in notifications or the status command will be hidden
* `notification_channel` - Channel ID where down/up notifications will be sent
* `retries` - How many retries to complete before flagging a server with down status
* `role_to_mention` - Role ID which will be tagged in down/up notifications
* `secs_between_ping` - How many seconds between each uptime check
* `timeout` - How many seconds before a ping or HTTP request should timeout

**NOTE: You must enable the [Message Content intent](https://support-dev.discord.com/hc/en-us/articles/4404772028055) in order to run v2.0+**.

## Running the bot

The bot can be run via the `bot.py` entrypoint, e.g. `python -m bot`.

## Docker Image
You can also run the Bot via Docker, simply run:

```bash
docker run --name discord-uptime -d -v $PWD/config.json:/app/config.json:ro -v $PWD/servers.json:/app/servers.json:ro --restart=on-failure ghcr.io/finlaysawyer/discord-uptime:master
```
You can also use a release tag (e.g. `v2.0.0`) instead of `master`.
With this configuration you still need to manually create `config.json` and `servers.json` for the Bot to start, please refer to the appropriate instructions if unsure.

## Servers Configuration
Servers should be setup similar to the examples already in `server.json`:
* There are three supported types: `http`, `tcp` and `ping`
```json
[
    {
      "name": "Google",
      "type": "http",
      "address": "google.com"
    },
    {
      "name": "Gmail SMTP",
      "type": "tcp",
      "address": "smtp.gmail.com:465",
    },
    {
      "name": "Cloudflare",
      "type": "ping",
      "address": "1.1.1.1"
    }
]
```

## Commands
> Default Prefix: >

* `ping <address> [pings]` - Pings an address once, or for the amount specified via pings and returns the delay in ms
* `http <address>` - Performs a HTTP request to the specified address and returns the response code
* `status` - Displays the status of all servers setup in `servers.json`

## Screenshots
> Status Command

![status](https://i.gyazo.com/6d5e0c4fbdb5ff52619d86eef827e369.png)

> Uptime & Downtime Notifications

![uptime](https://i.gyazo.com/803aebfcb3833ac8de7bd38e18378a29.png)

> Hide IP addresses

![hideips](https://i.gyazo.com/8596a75d33aa85716ba86f7e01621bb1.png)

## Contributing

Contributions are welcome. I use flake8 & black for linting and there is a [pre-commit](https://pre-commit.com/) config in the repo to run these tools against local commits.


CHANGELOG.md
# Changelog

## [1.5.0] - 2023-03-12

`v1.5.0` will be the last release in the v1 cycle. v2 is in development and will contain several improvements
over v1, such as the addition of slash commands, supporting the 'ping' check without root via icmplib, and more. Future
improvements such as persistent tracking of uptime/downtime in SQLite is also planned, but will likely come
after the initial v2 release.

### Added
- `pre-commit-ci` has been configured to autofix PRs that do not pass formatting checks
- **[BREAKING]** Python 3.7 has been dropped, as upstream Discord.py no longer supports it
- The message content intent has been added, which is a requirement from Discord to run the bot (thanks @dotneko)
- A Dockerfile has been added as an alternative way to run the bot (thanks @alsoGAMER)
- A workflow has been added that automatically pushes new image releases to [GHCR](https://github.com/finlaysawyer/discord-uptime/pkgs/container/discord-uptime)

### Changed
- Discord.py has been bumped to the latest 2.x version (thanks @dotneko)
- Modifications for newer async/await syntax (thanks @dotneko)
- Various typing, logging and spelling tweaks
- `aioping` is now used instead of `ping3` for the ping check

### Fixed
- IP addresses are now censored correctly when a port is added at the end

## [1.4.0] - 2021-09-25

### Added
- **[BREAKING]** Support for retrying status checks before sending an alert (thanks @tferreira!)

### Changed
- If multiple servers are down, the notification role will now only be mentioned once (thanks @ColdUnwanted!)
- `KeyError` will now be raised for config related errors

## [1.3.0] - 2021-06-24

### Added
- Support for tcp monitoring (thanks @tferreira!)
- Added a pre-commit config and flake8

### Changed
- Added support for building the bot against multiple Python versions in CI
- Minor documentation & typing tweaks

## [1.2.1] - 2021-03-22

### Changed
- User commands like `ping` will now escape mentions, preventing the bot from using tagging roles.

## [1.2.0] - 2021-03-20

### Added
- Option to hide IP addresses in commands and notifications (`hide_ips`) (#23).

### Changed
- `http_timeout` has been renamed to `timeout` and now also applies to ping requests.
- Up and down notifications will now use the server name in the Embed title instead of the address.

### Fixed
- `activity_name` now reads from the config correctly.
- `disable_help` now reads a boolean instead of a string.

## [1.1.0] - 2021-02-19

### Added
- Bandit for security linting
- User-friendly errors when a config is formatted incorrectly or missing values
- Configurable bot status in `config.json`:
  - `activity_type` - one of `playing, streaming, listening, watching`
  - `activity_name` - any string
- Default help command can now be disabled by setting `disable_help` to 'true'

### Changed
- Replaced flake8 with pylint and did some minor refactoring

## [1.0.1] - 2021-01-10

### Changed
- Updated the default config to reflect new `http_timeout` option

## [1.0.0] - 2021-01-10

### Added
- **Added support for tracking website uptime in montoring.**
- **Added http command for manual checking of website uptime.**
- Added more type hints and data types for function parameters
- Linting workflow for PRs and master releases
- More annotations for commands
- Added logging

### Changed
- **No longer need to reload the bot to update the config files.**
- Status command moved into the monitor cog.
- Status command and up/down notifications now show monitor type.
- Discord.py upgraded to 1.6 and aiohttp bumped to 3.7.3
- Main file renamed to `bot.py`

[1.0.0]: https://github.com/finlaysawyer/discord-uptime/compare/v0.0.1...v1.0.0
[1.0.1]: https://github.com/finlaysawyer/discord-uptime/compare/v1.0.0...v1.0.1
[1.1.0]: https://github.com/finlaysawyer/discord-uptime/compare/v1.0.1...v1.1.0
[1.2.0]: https://github.com/finlaysawyer/discord-uptime/compare/v1.1.1...v1.2.0
[1.2.1]: https://github.com/finlaysawyer/discord-uptime/compare/v1.2.0...v1.2.1
[1.3.0]: https://github.com/finlaysawyer/discord-uptime/compare/v1.2.1...v1.3.0
[1.4.0]: https://github.com/finlaysawyer/discord-uptime/compare/v1.3.0...v1.4.0
[1.5.0]: https://github.com/finlaysawyer/discord-uptime/compare/v1.4.0...v1.5.0


bot.py
from __future__ import annotations

import asyncio
import logging
import sys

import discord
from discord.ext import commands

from utils.config import get_config

logging.basicConfig(
    format="%(levelname)s | %(asctime)s | %(name)s | %(message)s",
    stream=sys.stdout,
    level=logging.INFO,
)
logger = logging.getLogger(__name__)

intents = discord.Intents.default()
intents.message_content = True

COGS = (
    "ping",
    # "tcp",
    # "http",
    # "monitor",
)


class DiscordUptime(commands.Bot):
    def __init__(self) -> None:
        super().__init__(
            command_prefix=get_config("prefix"),
            description="Bot to monitor uptime of services",
            reconnect=True,
            intents=intents,
            activity=discord.Activity(
                type=getattr(discord.ActivityType, get_config("activity_type").lower()),
                name=get_config("activity_name"),
            ),
            help_command=commands.DefaultHelpCommand()
            if not get_config("disable_help")
            else None,
        )

    async def setup_hook(self) -> None:
        for cog in COGS:
            await self.load_extension(f"cogs.{cog}")

        guild = discord.Object(id=get_config("guild_id"))
        self.tree.copy_global_to(guild=guild)
        await self.tree.sync(guild=guild)

    async def on_ready(self) -> None:
        logger.info(f"Logged in as {self.user}")

    async def on_command_error(
        self, ctx: commands.Context, error: commands.CommandError
    ) -> None:
        if isinstance(error, (commands.BadArgument, commands.MissingRequiredArgument)):
            await ctx.send(str(error))
        return None


async def main() -> None:
    await DiscordUptime().start(get_config("token"))


if __name__ == "__main__":
    asyncio.run(main())


cogs/ping.py
import asyncio

import aiohttp
from discord import Interaction, app_commands
from discord.ext import commands
from discord.utils import escape_mentions
from icmplib import ICMPLibError

from checks.ping import ping
from utils.config import get_config
from utils.embeds import CheckType, Status, generate_status_embed


class Ping(commands.Cog):
    def __init__(self, bot: commands.Bot) -> None:
        self.bot = bot

    @app_commands.command(description="Pings an address")
    async def ping(self, interaction: Interaction, address: str) -> None:
        """
        Pings an address via icmplib
        :param interaction: Discord interaction
        :param address: Address to ping
        :return: Embed message containing status of service
        """
        await interaction.response.send_message("Pinging...")
        address = escape_mentions(address)

        try:
            ping_request = await ping(address)
        except ICMPLibError as err:
            status = Status.DOWN
            embed_fields: list[dict] = [{"name": "Reason", "value": str(err)}]
        else:
            # If no packets could be sent or packet loss was encountered,
            # consider the service down
            if ping_request.packet_loss > 0 or ping_request.packets_sent == 0:
                status = Status.DOWN
            else:
                status = Status.UP

            embed_fields = [
                {
                    "name": "Packets received",
                    "value": f"{ping_request.packets_sent} / {ping_request.packets_received}",
                    "inline": False,
                },
                {
                    "name": "Packet loss",
                    "value": f"{ping_request.packet_loss:.0%}",
                    "inline": False,
                },
                {
                    "name": "Round-trip time",
                    "value": f"Minimum: {ping_request.min_rtt}ms"
                    f"\nAverage: {ping_request.avg_rtt}ms"
                    f"\nMaximum: {ping_request.max_rtt}ms",
                    "inline": False,
                },
            ]

        await interaction.edit_original_response(
            content=None,
            embed=generate_status_embed(
                status,
                CheckType.PING,
                address,
                fields=embed_fields,
            ),
        )

    @app_commands.command(description="Checks a TCP port")
    async def tcp(self, interaction: Interaction, address: str, port: int) -> None:
        """
        Checks if a TCP port on a remote host is open for connections
        :param interaction: Discord Interaction
        :param address: Address of host
        :param port: Port to connect to
        :return: Delay in milliseconds or error
        """
        timeout = get_config("timeout")
        address = escape_mentions(address)

        conn = asyncio.open_connection(address, port)
        try:
            reader, writer = await asyncio.wait_for(conn, timeout)
            await interaction.response.send_message(
                f"Connection established on {address}:{port}"
            )
            writer.close()
            await writer.wait_closed()
        except asyncio.TimeoutError:
            await interaction.response.send_message(
                f"Request timed out after {timeout} seconds"
            )
        except ConnectionRefusedError:
            await interaction.response.send_message(
                f"Could not establish a connection to {address}:{port}"
            )

    @app_commands.command(description="Performs an HTTP request")
    async def http(self, interaction: Interaction, address: str) -> None:
        """
        Performs an HTTP request to the specified address
        :param interaction: Discord Interaction
        :param address: Address to make request to
        :return: HTTP status code
        """
        if not address.startswith("http"):
            address = f"https://{address}"

        timeout = get_config("timeout")
        address = escape_mentions(address)

        async with aiohttp.ClientSession(
            timeout=aiohttp.ClientTimeout(total=timeout)
        ) as session:
            try:
                async with session.get(address) as res:
                    await interaction.response.send_message(
                        f"Received response code: {res.status} ({res.reason})"
                    )
            except asyncio.TimeoutError:
                await interaction.response.send_message(
                    f"Request timed out after {timeout} seconds"
                )
            except aiohttp.ClientError:
                await interaction.response.send_message(
                    f"Could not establish a connection to {address}"
                )


async def setup(bot: commands.Bot) -> None:
    await bot.add_cog(Ping(bot))


cogs/status.py
from datetime import timedelta

import discord
from discord.ext import commands

from cogs import monitor
from utils import config as cfg


class Status(commands.Cog):

    def __init__(self, bot):
        self.bot = bot

    @commands.command(brief="Checks status of servers - status")
    async def status(self, ctx):
        """
        :param ctx:
        :return: Embed of the status of each server
        """
        embed = discord.Embed(title="**Monitor Status**", color=16711680 if monitor.currently_down else 65287)

        for i in cfg.servers:
            if i['address'] in monitor.currently_down:
                downtime = str(timedelta(seconds=monitor.currently_down[i['address']]))
                embed.add_field(name=i['name'], value=f":red_circle: {i['address']} ({downtime})", inline=False)
            else:
                embed.add_field(name=i['name'], value=f":green_circle: {i['address']}", inline=False)

        await ctx.send(embed=embed)


def setup(bot):
    bot.add_cog(Status(bot))


utils/config.py
import json
from typing import Any, Iterable

GREEN = 65287
RED = 16711680


def load_config_file(file: str) -> dict:
    """Attempts to load a json config file"""
    with open(file) as config:
        try:
            return json.load(config)
        except json.decoder.JSONDecodeError as err:
            raise KeyError(
                f"Couldn't load {config}: it is formatted incorrectly "
                f"on line {err.lineno} column {err.colno}"
            ) from err


def get_servers() -> Iterable:
    """Reads all specified servers specified in servers.json."""
    return load_config_file("servers.json")


def get_server_name(address: str) -> str:
    """Retrieves a stored server name from an address."""
    file = load_config_file("servers.json")

    for server in file:
        if server["address"] == address:
            return server["name"]

    raise KeyError(f"Could not fetch a server name from {address}")


def get_config(item: str) -> Any:
    """Retrieves the configuration value specified."""
    file = load_config_file("config.json")

    value = file.get(item)

    if value is None:
        raise KeyError(f"Your config is out of date! Missing a value for {item}")
    return value


utils/embeds.py
from __future__ import annotations

import re
from datetime import datetime
from enum import Enum
from typing import Any

import discord

from utils.config import GREEN, RED, get_config


def hide_ips(var: str) -> str:
    """Replaces IP addresses and ports in a given string"""
    return re.sub(r"[0-9]+(?:\.[0-9]+){3}(:[0-9]+)?", "*hidden*", str(var))


class Embed(discord.Embed):
    def add_field(self, *, name: Any, value: Any, inline: bool = True) -> Embed:
        """Override the add_field method in the original embed to hide addresses"""

        if get_config("hide_ips"):
            name = hide_ips(name)
            value = hide_ips(value)

        field = {"inline": inline, "name": str(name), "value": str(value)}

        try:
            self._fields.append(field)
        except AttributeError:
            self._fields = [field]

        return self


class Status(Enum):
    UP = 1
    DOWN = 2


class CheckType(Enum):
    PING = 1
    HTTP = 2
    TCP = 3


def generate_status_embed(
    status: Status,
    check_type: CheckType,
    request_address: str,
    fields: list[dict] | None = None,
) -> Embed:
    """
    Generates an embed for a request.
    :param status: Indicates whether the request was successful or failed
    :param check_type: The type of request made
    :param request_address: The address the request was mde to
    :param fields: A list of embed fields in the form of dictionaries
    :return: Generated embed
    """
    colour = GREEN if status == Status.UP else RED
    emoji = ":green_circle:" if status == Status.UP else ":red_circle:"
    status_text = "succeeded" if status == Status.UP else "failed"

    embed = Embed(
        title=f"**{emoji} {check_type.name.capitalize()} request to {request_address.lower()} {status_text}**",
        timestamp=datetime.now(),
        colour=colour,
    )

    if fields:
        for field in fields:
            embed.add_field(**field)

    return embed


discord-uptime.py
from discord import Intents
from discord.ext import commands
from utils import config as cfg
import os

bot = commands.Bot(command_prefix=cfg.config['prefix'])


class DiscordUptime(commands.Bot):
    def __init__(self):
        super().__init__(command_prefix=cfg.config['prefix'],
                         description='Bot to monitor uptime of services',
                         reconnect=True,
                         intents=Intents.default())
        self.bot = bot

    async def on_ready(self):
        print(f"Logged in as {self.user}")

        for filename in os.listdir('./cogs'):
            if filename.endswith('.py'):
                self.load_extension(f'cogs.{filename[:-3]}')

    async def on_command_error(self, ctx, error):
        if isinstance(error, (commands.CommandNotFound, commands.BadArgument, commands.MissingRequiredArgument)):
            return await ctx.send(error)
        else:
            return


DiscordUptime().run(cfg.config["token"])


cogs/monitor.py
from __future__ import annotations

import asyncio
import logging
from datetime import timedelta

import aiohttp
import aioping
import discord
from discord.ext import commands, tasks

from utils import embeds
from utils.config import get_config, get_server_name, get_servers

logger = logging.getLogger(__name__)


class Monitor(commands.Cog):
    def __init__(self, bot: commands.Bot):
        self.bot = bot
        self.retry_count: dict[str, int] = {}
        self.currently_down: dict[str, int] = {}
        self.monitor_uptime.start()
        self.need_to_mention = False
        self.currently_checking = False
        logger.info("Monitor initialized")

    async def cog_unload(self) -> None:
        self.monitor_uptime.cancel()

    def needs_retry(self, server: dict) -> bool:
        retry_count = self.retry_count.get(server["address"], 0)
        if retry_count < get_config("retries"):
            self.retry_count[server["address"]] = retry_count + 1
            return True
        return False

    async def notify_down(
        self, server: dict, channel: discord.TextChannel, reason: str | None
    ) -> None:
        """
        Sends an embed to indicate a service is offline
        :param server: Server object to extract data from
        :param channel: Channel to send the notification to
        :param reason: Reason why the service is down
        """
        if self.needs_retry(server):
            return

        if server["address"] not in self.currently_down:
            logger.info(f'Server {server["address"]} went down')

            self.currently_down.update({server["address"]: 0})
            embed = embeds.Embed(
                title=f"**:red_circle: {get_server_name(server['address'])} is down!**",
                color=16711680,
            )
            embed.add_field(name="Address", value=server["address"], inline=False)
            embed.add_field(name="Type", value=server["type"], inline=False)
            embed.add_field(name="Reason", value=reason, inline=False)

            try:
                await channel.send(embed=embed)
            except discord.Forbidden:
                logger.error(
                    "Down notification could not be sent. "
                    "Ensure the bot has permission to sent messages to the specified channel."
                )
            except Exception as e:
                logger.exception(e)

            if self.need_to_mention is False:
                self.need_to_mention = True
        else:
            self.currently_down[server["address"]] = self.currently_down.get(
                server["address"], 0
            ) + get_config("secs_between_ping")

    async def notify_up(self, server: dict, channel: discord.TextChannel) -> None:
        """
        Sends an embed to indicate a service is online
        :param server: Server object to extract data from
        :param channel: Channel to send the notification to
        """
        if server["address"] in self.currently_down:
            logger.info(f'Server {server["address"]} is back online')

            embed = embeds.Embed(
                title=f"**:green_circle: {get_server_name(server['address'])} is up!**",
                color=65287,
            )
            embed.add_field(name="Address", value=server["address"], inline=False)
            embed.add_field(name="Type", value=server["type"], inline=False)
            embed.add_field(
                name="Downtime",
                value=str(timedelta(seconds=self.currently_down[server["address"]])),
                inline=False,
            )

            try:
                await channel.send(embed=embed)
            except discord.Forbidden:
                logger.error(
                    "Up notification could not be sent. "
                    "Ensure the bot has permission to sent messages to the specified channel."
                )
            except Exception as e:
                logger.exception(e)

            if self.need_to_mention is False:
                self.need_to_mention = True

            self.currently_down.pop(server["address"])
            try:
                if self.retry_count is not {}:
                    self.retry_count.pop(server["address"])
            except Exception as e:
                logger.exception(e)
                logger.debug(f"self.retry_count={self.retry_count}")

    @tasks.loop(seconds=get_config("secs_between_ping"))
    async def monitor_uptime(self) -> None:
        """Checks the status of each server and sends up/down notifications"""
        await self.bot.wait_until_ready()

        channel: discord.TextChannel | None = self.bot.get_channel(get_config("notification_channel"))  # type: ignore

        if not channel:
            logger.debug(
                "Skipping scheduled monitoring job, invalid `notification_channel` configured..."
            )
            return

        timeout = get_config("timeout")

        # Make sure the need to mention is set to false everytime this function is ran
        self.need_to_mention = False

        for i in get_servers():
            self.currently_checking = True

            if i["type"] == "ping":
                try:
                    await aioping.ping(i["address"], timeout=timeout)
                except Exception as err:
                    await self.notify_down(i, channel, str(err))
                else:
                    await self.notify_up(i, channel)
            elif i["type"] == "tcp":
                host, port = i["address"].split(":")
                conn = asyncio.open_connection(host, port)
                try:
                    reader, writer = await asyncio.wait_for(conn, timeout)
                    writer.close()
                    await writer.wait_closed()
                except asyncio.TimeoutError:
                    await self.notify_down(i, channel, "Timed out")
                except ConnectionRefusedError:
                    await self.notify_down(i, channel, "Connection failed")
                else:
                    await self.notify_up(i, channel)
            else:
                address = i["address"]

                if not address.startswith("http"):
                    address = f"https://{address}"

                async with aiohttp.ClientSession(
                    timeout=aiohttp.ClientTimeout(total=timeout)
                ) as session:
                    try:
                        async with session.get(address) as res:
                            if res.ok:
                                await self.notify_up(i, channel)
                            else:
                                await self.notify_down(i, channel, res.reason)
                    except asyncio.TimeoutError:
                        await self.notify_down(i, channel, "Timed out")
                    except aiohttp.ClientError:
                        await self.notify_down(i, channel, "Connection failed")

        self.currently_checking = False
        role_to_mention = get_config("role_to_mention")

        if self.need_to_mention is True and role_to_mention != 0:
            try:
                await channel.send(f"<@&{role_to_mention}>", delete_after=3)
            except discord.Forbidden:
                logger.error(
                    "Mention could not be sent. "
                    "Ensure the bot has permission to sent messages to the specified channel."
                )
            except Exception as e:
                logger.exception(e)

    @commands.command(brief="Checks status of servers being monitored", usage="status")
    async def status(self, ctx: commands.Context) -> None:
        """Make this function asleep if the monitor is currently checking"""
        while self.currently_checking:
            await asyncio.sleep(0.1)

        """Returns an embed showing the status of each monitored server"""
        embed = embeds.Embed(
            title="**Monitor Status**", color=16711680 if self.currently_down else 65287
        )

        for i in get_servers():
            if i["address"] in self.currently_down:
                downtime = str(timedelta(seconds=self.currently_down[i["address"]]))
                embed.add_field(
                    name=f"{i['name']} ({i['type']})",
                    value=f":red_circle: {i['address']} ({downtime})",
                    inline=False,
                )
            else:
                embed.add_field(
                    name=f"{i['name']} ({i['type']})",
                    value=f":green_circle: {i['address']}",
                    inline=False,
                )

        await ctx.send(embed=embed)


async def setup(bot: commands.Bot) -> None:
    await bot.add_cog(Monitor(bot))
